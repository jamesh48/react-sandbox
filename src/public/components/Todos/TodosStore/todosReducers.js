// import {  } from './Actions.js';

const combineReducers = (slices) => (state, action) => Object.keys(slices).reduce((acc, prop) => ({
  ...acc,
  [prop]: slices[prop](acc[prop], action),
}),
  state
);

const reduceReducers = (...reducers) => {
  return (state, action) => reducers.reduce((acc, nextReducer) => nextReducer(acc, action), state)
};

const todos = (state = [], { type, payload }) => {
  switch (type) {
    case 'ADD TODO':
      return state.concat(payload);
    default:
      return state;
  };
};

const editedTodoValue = (state = '', {type, payload}) => {
  switch (type) {
    case 'EDIT TODO VALUE':
        return payload;
    default:
      return state;
  };
};

export default combineReducers({ todos, editedTodoValue });