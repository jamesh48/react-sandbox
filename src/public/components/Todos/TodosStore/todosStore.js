import React from 'react';
import CombinedReducers from './todosReducers.js';
const TodosStoreContext = React.createContext();
const initialState = {
  todos: [],
  editedTodoValue: ''
};
const TodosStoreProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(CombinedReducers,
    initialState);
  const store = React.useMemo(() => [state, dispatch], [state]);
  return (
    <TodosStoreContext.Provider value={store}>{children}</TodosStoreContext.Provider>
  );
};

const useTodosContext = () => {
  const context = React.useContext(TodosStoreContext);
  if (context === undefined) {
    throw new Error('useTodosContext must be used within a TodosProvider')
  }
  return context;
};
export { useTodosContext, TodosStoreProvider };