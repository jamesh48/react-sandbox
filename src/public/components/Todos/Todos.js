import React from 'react'
import globalStyles from 'GlobalSCSS';
import { useTodosContext } from 'TodosStore';
const { componentContainer } = globalStyles;

export default () => {
  const [{ todos, editedTodoValue }, todosDispatch] = useTodosContext();


  React.useEffect(() => {
    todosDispatch({ type: 'EDIT TODO VALUE', payload: '' });
  }, [todos]);

  return (
    <div className={componentContainer} id='todos-root'>
      <div>Todos: {todos.join(', ')}</div>
      <form onSubmit={() => {
        event.preventDefault();
        todosDispatch({ type: 'ADD TODO', payload: editedTodoValue })
      }}>
        <input type='text' value={editedTodoValue} onChange={() => {
          todosDispatch({ type: 'EDIT TODO VALUE', payload: event.target.value })
        }}></input>
        <button type='submit'>Add to Todos</button>
      </form>
    </div>
  );
};