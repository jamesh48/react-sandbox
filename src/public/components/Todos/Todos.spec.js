import React from 'react';
import Todos from './Todos.js';
import { TodosStoreProvider } from 'TodosStore';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('Todos Component', () => {
  it('Renders the Todos Component', () => {
    const wrapper = mount(
      <TodosStoreProvider>
        <Todos />
      </TodosStoreProvider>
    );
    expect(wrapper.find('#todos-root').length).toEqual(1);
  })

  test('todos snapshot renders', () => {
    const component = renderer.create();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});