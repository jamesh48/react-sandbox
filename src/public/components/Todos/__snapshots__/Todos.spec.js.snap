// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Component todos snapshot renders 1`] = `null`;
